### PR TITLE
Read/write settings using core api

### LINKED ISSUE #1686: Extend LuigiCore API to read/write settings
These functions should be member of `configuration`.

### PR DESCRIPTION
**Pull request by JohannesDoberer**
**Description**

Changes proposed in this pull request:

- Read and write settings to localStorage
-  OR if a function `readFromCustomStorage`/`writeToCustomStorage` is specified in `settings` section of the luigi config these functions will be called instead of calling localStorage 

**Related issue(s)**
Resolves #1686

### CODE REVIEW

File: test/e2e-test-application/src/luigi-config/extended/settings.js [hunk 1]
```diff
@@ -55,6 +55,56 @@ class Settings {
...
     };
   }; */
+
+  userSettings = {
+    userSettingsNavigationEntry: {
+      label: 'My Settings',
+      icon: 'settings'
+    },
+    writeToCustomStorage: (key, obj) => {
+      sessionStorage.setItem(key, JSON.stringify(obj));
```
Comment (stanleychh) [inline test/e2e-test-application/src/luigi-config/extended/settings.js:65]: Just curious, shall we use `localStorage` or `sessionStorage`?

**Review by stanleychh — COMMENTED** [review]
Good job. Overall looks good.

File: core/test/services/config.spec.js [hunk 2]
```diff
@@ -124,4 +126,56 @@ describe('Config', () => {
...
+    afterEach(() => {
+      sinon.restore();
+      sinon.reset();
+    });
+    const key = 'myStorageKey'
+    const userSettingsObj = {
+      userSettings: {
+        some: "object"
+      }
+    }
```
[the remark this answers is not in the document]
Comment (JohannesDoberer) [reply core/test/services/config.spec.js]: 👍  you are right.

File: test/e2e-test-application/src/luigi-config/extended/settings.js (hunk 1 above)
Comment (JohannesDoberer) [reply test/e2e-test-application/src/luigi-config/extended/settings.js]: These functions are there if you don't want to use localstorage. You can implement these functions and implement your own logic to either use session storage or call some api to store data there.
If these functions are not implemented the localstorage will be used.

**Review by stanleychh — APPROVED** [review]

File: core/src/core-api/config.js [hunk 3]
```diff
@@ -232,6 +232,49 @@ class LuigiConfig {
       container.removeChild(container.lastChild);
     }
   }
+
+  /**
+   * Reads the user settings object from the localstorage. It is possible to choose a custom storage type by implementing the `settings.userSetting.readFromCustomStorage` function.
```
Comment (alexandra-simeonova) [inline core/src/core-api/config.js:237]:
```suggestion
   * Reads the user settings object from the **localStorage**. It is possible to choose a custom storage type by implementing the `settings.userSetting.readFromCustomStorage` function.
```

File: core/src/core-api/config.js [hunk 4]
```diff
@@ -232,6 +232,49 @@ class LuigiConfig {
...
+   * @memberof Configuration
+   * @param {string} key used to identify the item in the storage.
+   * @param {string} property the object traversal path where `readFromCustomStorage` function is implemented. Optional and only needed if there is a custom implementation of `readFromCustomStorage` function in the settings section of the luigi config to read a custom storage.
```
Comment (alexandra-simeonova) [inline core/src/core-api/config.js:240]:
```suggestion
   * @param {string} property the object traversal path where the `readFromCustomStorage` function is implemented. Optional and only needed if there is a custom implementation of a `readFromCustomStorage` function in the settings section of the Luigi configuration to read a custom storage.
```

File: core/src/core-api/config.js [hunk 5]
```diff
@@ -232,6 +232,49 @@ class LuigiConfig {
...
+      if (settings && GenericHelpers.isFunction(settings.readFromCustomStorage)) {
+        return settings.readFromCustomStorage(key);
+      }
+    } else {
+      return JSON.parse(localStorage.getItem(key));
+    }
+  }
+
+  /**
+   * Writes the user settings object to the localstorage. It is possible to choose a custom storage type by implementing the `settings.userSetting.writeToCustomStorage` function.
```
Comment (alexandra-simeonova) [inline core/src/core-api/config.js:258]:
```suggestion
   * Writes the user settings object to the **localStorage**. It is possible to choose a custom storage type by implementing the `settings.userSetting.writeToCustomStorage` function.
```

File: core/src/core-api/config.js [hunk 6]
```diff
@@ -232,6 +232,49 @@ class LuigiConfig {
...
+   * @memberof Configuration
+   * @param {string} key used to identify the item in the storage.
+   * @param {Object} settingsObject to store in the storage.
+   * @param {string} property the object traversal path where `writeToCustomStorage` function is implemented. Optional and only needed if there is a custom implementation of `writeToCustomStorage` function in the settings section of the luigi config to read a custom storage.
```
Comment (alexandra-simeonova) [inline core/src/core-api/config.js:262]:
```suggestion
   * @param {string} property the object traversal path where the `writeToCustomStorage` function is implemented. Optional and only needed if there is a custom implementation of a `writeToCustomStorage` function in the settings section of the Luigi configuration to read a custom storage.
```

File: core/src/core-api/config.js [hunk 7]
```diff
@@ -232,6 +232,49 @@ class LuigiConfig {
...
   }
+
+  /**
+   * Reads the user settings object from the localstorage. It is possible to choose a custom storage type by implementing the `settings.userSetting.readFromCustomStorage` function.
+   * @memberof Configuration
+   * @param {string} key used to identify the item in the storage.
+   * @param {string} property the object traversal path where `readFromCustomStorage` function is implemented. Optional and only needed if there is a custom implementation of `readFromCustomStorage` function in the settings section of the luigi config to read a custom storage.
+   * @example
+   * Luigi.writeUserSettings('key');
+   * Luigi.writeUserSettings('key', 'settings.userSettings');
```
Comment (UlianaMunich) [inline core/src/core-api/config.js:243]:
```suggestion
   * Luigi.readSettingsFromStorage('key');
   * Luigi.readSettingsFromStorage('key', 'settings.userSettings');
```

File: docs/luigi-core-api.md [hunk 8]
```diff
@@ -172,6 +172,47 @@ Luigi.unload()
...
+##### Parameters
+
+-   `key` **[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** used to identify the item in the storage.
+-   `property` **[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** the object traversal path where `readFromCustomStorage` function is implemented. Optional and only needed if there is a custom implementation of `readFromCustomStorage` function in the settings section of the luigi config to read a custom storage.
+
+##### Examples
+
+```javascript
+Luigi.writeUserSettings('key');
+Luigi.writeUserSettings('key', 'settings.userSettings');
```
Comment (UlianaMunich) [inline docs/luigi-core-api.md:188]:
```suggestion
Luigi.readFromCustomStorage('key');
Luigi.readFromCustomStorage('key', 'settings.userSettings');
```

**Review by UlianaMunich — CHANGES_REQUESTED** [review]
Please check my suggestions in docu

File: core/src/core-api/config.js (hunk 7 above)
Comment (JohannesDoberer) [reply core/src/core-api/config.js:243]: Done

File: docs/luigi-core-api.md (hunk 8 above)
Comment (JohannesDoberer) [reply docs/luigi-core-api.md]: Done, it is named `writeSettingsToStorage`. Thx for finding, hope name is ok.

**Review by UlianaMunich — APPROVED** [review]
LGTM!

File: core/src/core-api/config.js [hunk 9]
```diff
@@ -232,6 +233,46 @@ class LuigiConfig {
...
+   * @since NEXTRELEASE
+   */
+  async readUserSettings() {
+    const userSettings = await this.getConfigValueAsync('settings.userSettings');
+    if (userSettings && GenericHelpers.isFunction(userSettings.readUserSettings)) {
+      return userSettings.readUserSettings();
+    }
+    else {
+      return JSON.parse(localStorage.getItem(this.USER_SETTINGS_KEY));
+    }
```
Comment (stanleychh) [inline core/src/core-api/config.js:253]:
```suggestion
    return JSON.parse(localStorage.getItem(this.USER_SETTINGS_KEY));
```

File: core/src/core-api/config.js [hunk 10]
```diff
@@ -232,6 +233,44 @@ class LuigiConfig {
       container.removeChild(container.lastChild);
     }
   }
+
+  /**
+   * Reads the user settings object.
+   * You can choose a custom storage to read the user settings by implementing the `userSetting.writeUserSettings` function in the settings section of the Luigi configuration.
```
Comment (stanleychh) [inline core/src/core-api/config.js:239]:
```suggestion
   * You can choose a custom storage to read the user settings by implementing the `userSetting.readUserSettings` function in the settings section of the Luigi configuration.
```

File: core/src/core-api/config.js [hunk 11]
```diff
@@ -232,6 +233,44 @@ class LuigiConfig {
...
+    const userSettings = await this.getConfigValueAsync('settings.userSettings');
+    if (userSettings && GenericHelpers.isFunction(userSettings.readUserSettings)) {
+      return userSettings.readUserSettings();
+    }
+    return JSON.parse(localStorage.getItem(this.USER_SETTINGS_KEY));
+  }
+
+  /**
+   * Reads the user settings object.
+   * You can choose a custom storage to write the user settings by implementing the `userSetting.writeUserSettings` function in the settings section of the Luigi configuration
```
Comment (stanleychh) [inline core/src/core-api/config.js:256]:
```suggestion
   * You can choose a custom storage to write the user settings by implementing the `userSetting.storeUserSettings` function in the settings section of the Luigi configuration
```

**Review by stanleychh — CHANGES_REQUESTED** [review]
Two more minor comments

File: core/src/core-api/config.js (hunk 11 above)
Comment (JohannesDoberer) [reply core/src/core-api/config.js]: thx 👍

File: core/src/core-api/config.js (hunk 10 above)
Comment (JohannesDoberer) [reply core/src/core-api/config.js]: thx 👍

**Review by alexandra-simeonova — APPROVED** [review]

**Review by stanleychh — APPROVED** [review]

File: core/test/services/config.spec.js [hunk 12]
```diff
@@ -124,4 +126,55 @@ describe('Config', () => {
...
+      };
+    });
+    afterEach(() => {
+      sinon.restore();
+      sinon.reset();
+    });
+    const key = 'myStorageKey';
+    const userSettingsObj = {
+      userSettings: {
+        userSettingGroups: {}
```
Comment (UlianaMunich) [inline core/test/services/config.spec.js:144]: to be consistent I would suggest to rename it to `userSettingsGroups`, we use everywhere the plural: userSettings

**Pull request merged**